From 6cd50eba590f292c13b060183a23ff0c1807eb9b Mon Sep 17 00:00:00 2001
From: Joe Gordon <jogo@cloudscaling.com>
Date: Tue, 5 Feb 2013 18:24:21 -0800
Subject: [PATCH 3/3] Update nova/compute/api to handle instance as dict

When nova.dp.api returns primitives instead of sqlalchemy objects
instance objects will be dictionaries, which cannot be added to sets
('dict' is an unhashable type)

Partially implements bp db-api-cleanup

Change-Id: I3683d1c54ae627d9cabfd569a7fd304c47c5d5b7
---
 nova/compute/api.py | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/nova/compute/api.py b/nova/compute/api.py
index 00eed2f..4e482d6 100644
--- a/nova/compute/api.py
+++ b/nova/compute/api.py
@@ -2534,13 +2534,14 @@ class SecurityGroupAPI(base.Base):
             security_groups.add(security_group)
 
         # ..then we find the instances that are members of these groups..
-        instances = set()
+        instances = {}
         for security_group in security_groups:
             for instance in security_group['instances']:
-                instances.add(instance)
+                if instance['uuid'] not in instances:
+                    instances[instance['uuid']] = instance
 
         # ..then we send a request to refresh the rules for each instance.
-        for instance in instances:
+        for instance in instances.values():
             if instance['host']:
                 self.security_group_rpcapi.refresh_instance_security_rules(
                         context, instance['host'], instance)
-- 
1.8.1.4

